kratos/geometries/nurbs_shape_function_utilities/nurbs_surface_refinement_utilities.cpp

```diff
@@ -0,0 +1,329 @@
+//    |  /           |
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:         BSD License
+//                   Kratos default license: kratos/license.txt
+//
+
+// System includes
+
+// External includes
+
+// Project includes
+#include "nurbs_surface_refinement_utilities.h"
+
+namespace Kratos {
+
+    ///@name Operations
+    ///@{
+
+    void NurbsSurfaceRefinementUtilities::KnotRefinementU(
+        NurbsSurfaceGeometryType& rGeometry,
+        std::vector<double>& rKnotsUToInsert,
+        PointerVector<NodeType>& rPointsRefined,
+        Vector& rKnotsURefined,
+        Vector& rWeightsRefined)
+    {
+        std::sort(rKnotsUToInsert.begin(), rKnotsUToInsert.end());
+
+        const SizeType nb_knots_u_to_insert = rKnotsUToInsert.size();
+
+        const SizeType degree_u = rGeometry.PolynomialDegree(0);
+
+        const SizeType nb_cp_u_old = rGeometry.PointsNumberInDirection(0);
+        const SizeType nb_cp_v = rGeometry.PointsNumberInDirection(1);
+
+        const Vector& knots_u_old = rGeometry.KnotsU();
+
+        const Vector& weights_old = rGeometry.Weights();
+
+        const SizeType nb_knots_u_old = knots_u_old.size();
+
+        const SizeType a = NurbsUtilities::GetUpperSpan(degree_u, knots_u_old, *knots_u_old.begin());
+        const SizeType b = NurbsUtilities::GetUpperSpan(degree_u, knots_u_old, *knots_u_old.end());
```

Comment:
AFAIK `.end()` is pointing to `vector[n + 1]`. Maybe use `front` and `back` ([see](https://github.com/oberbichler/ANurbs/blob/e508daa256a292850e25ca48bc74ee8cfbe4efd2/include/anurbs/Algorithm/KnotRefinement.h#L37-L38)).

Reply:
thanks for that hint. Indeed `.end()` should not be correct. The UBLAS vector however does not has `.back()`